Reland of a bunch of pattern matcher + indexing improvements

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112476



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

re: the optimization landed in this PR, see https://github.com/pytorch/pytorch/pull/112102#issuecomment-1780265536

Chillee committed:
Reland of a bunch of pattern matcher + indexing fixes

[ghstack-poisoned]

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Chillee commented:
@chillee has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50826866).

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D50826866](https://our.internmc.facebook.com/intern/diff/D50826866)

[ghstack-poisoned]

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D50826866](https://our.internmc.facebook.com/intern/diff/D50826866)

[ghstack-poisoned]

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D50826866](https://our.internmc.facebook.com/intern/diff/D50826866)

[ghstack-poisoned]

oulgen commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -657,7 +666,7 @@ def can_replace(node, mutated_arg):
                 return False
 
-            # Check for any uses other than current node and copy_ epilogue
+            # # Check for any uses other than current node and copy_ epilogue
```
you commented your comment

oulgen commented on torch/_inductor/lowering.py:
```diff
@@ -2845,6 +2846,7 @@ def index_impl(x, indices, check):
         indices_loaders,
         indexed_size,
         x_loader,
+        check=check,
```
did we not rename this?

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D50826866](https://our.internmc.facebook.com/intern/diff/D50826866)

[ghstack-poisoned]

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D50826866](https://our.internmc.facebook.com/intern/diff/D50826866)

[ghstack-poisoned]

Chillee committed:
Update on "Reland of a bunch of pattern matcher + indexing fixes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D50826866](https://our.internmc.facebook.com/intern/diff/D50826866)

[ghstack-poisoned]

Chillee commented:
@pytorchbot merge

Chillee commented:
@pytorchbot merge
